fl4health/model_bases/autoencoders_base.py

```diff
@@ -0,0 +1,144 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class AutoEncoderType(Enum):
+    BASIC_AE = "BASIC_AE"
+    VARIATIONAL_AE = "VARIATIONAL_AE"
+    CONDITIONAL_VAE = "CONDITIONAL_VAE"
+
+
+class AbstractAE(nn.Module, ABC):
+    """The base class for all Encoder-Decoder based models.
+    All we need to define such model is the type of the model, and the structure
+    of the encoder and the decoder modules. This type of model should have the capability
+    to encode data using the encoder module and decode the output of the encoder using the decoder module.
+    """
+
+    def __init__(
+        self,
+        model_type: AutoEncoderType,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        super().__init__()
+        self.model_type = model_type
+        self.encoder = encoder
+        self.decoder = decoder
+
+    @abstractmethod
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class VariationalAE(AbstractAE):
+    """Variational Auto-Encoder model base class."""
+
+    def __init__(
+        self,
+        model_type: AutoEncoderType,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        """Initializes the VariationalAE class.
+
+        Args:
+            model_type (AutoEncoderType): The model type should be VARIATIONAL_AE.
+            encoder (nn.Module): Encoder module defined by the user.
+            decoder (nn.Module): Decoder module defined by the user.
+        """
+        super().__init__(model_type, encoder, decoder)
+
+    def encode(self, input: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        mu, logvar = self.encoder(input)
+        return mu, logvar
+
+    def decode(self, latent_vector: torch.Tensor) -> torch.Tensor:
+        output = self.decoder(latent_vector)
+        return output
+
+    def sampling(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        std = torch.exp(0.5 * logvar)
+        eps = torch.randn_like(std)
+        return eps.mul(std).add_(mu)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        mu, logvar = self.encode(input)
+        z = self.sampling(mu, logvar)
+        output = self.decode(z)
+        return torch.cat((logvar, mu, output), dim=1)
```

Comment:
Following from my [comment here](https://github.com/VectorInstitute/FL4Health/pull/82/files#r1432752549), I think concatenating the output (mean, var and output) is the best solution for now in terms of simplicity. However, I think this means we are making the assumption that inputs and outputs of the autoencoder will be flat which limits the architectures that can be used (ex. using conv net on image data). 

This may be a candidate for a follow up PR, but have you thought of anyway to avoid having this constraint? I was thinking maybe storing the spatial dims of samples in forward so we can do the reshaping. Then we would only require that the latent space by of a single dimension. This is still a bit hacky though and I havent fully thought the idea through. Feel free to add a clickup in ticket if you agree :)

Reply:
That is a very good point! I had some examples with CNNs but later when I implemented the `autoencoder_base` I only used this MLP structure, however, we definitely need to support all the architectures. Your idea for saving spatial dims makes sense. Somehow, a similar thing is happening here by passing the `latent_dim` to the `VAE_loss` for unpacking the output, mu, and logvar. So we are assuming the model is packing the output according to their latent dimension and then `VAE_loss` needs to unpack that the same way. 
 I am passing the `latent_dim` directly to `VAE_loss` to avoid importing the `autoencoder_base` model in the `loss.py`. By relaxing this assumption, one of the things that we can do is to define a packing and unpacking function in the model (AbstractAE), then use the unpacking function with the saved spatial dims anywhere that we need to unpack. I also haven't fully thought about this, but we can try this idea as it makes sense to me.

Reply:
Just want to make a comment here that I faced a similar issue in my federated PCA PR. In my case, in order to perform PCA, I had to assume the data points were flattened already. My solution there was to add a `maybe_reshape` method which is called in `forward` and flattens the input as needed. I wasn't sure if this was the best approach though.

Reply:
Maybe instead of directly calling `self.criterion` we can add a function to basic client to also handle dictionaries?  so the input to loss would be much more dynamic.